comptime.md

```diff
@@ -0,0 +1,125 @@
+- No way to invoke arbitrary V functions
+- No way to choose function at compile time: 
+```v
+$if flag_a {
+    fn foo() {
+        println('bar')
+    }
+} $else {
+    fn foo() {
+        println('foo')
+    }
+}
+```
+^ Compiler cannot handle this code 
+
+- It limits our generics and compile-time reflection. With proper comptime it would be possible to generate complex serialization/deserialization functions without compiler support. 
+- Implementing proper comptime in fact would simplify our compiler pipelines, at the moment there is a lot of edge-cases for handling current limited comptime. 
+- Proper metaprogramming is must have in every modern language 
+
+
+# Guide-level explanation 
+
+With this RFC the main change is that all expressions and statements prefixed with `$` will be treated as compile-time and all of them can be executed *before* codegen. Current compile-time is quite limited and allows only a few statements (`$if`,`$for`) and only specific calls.
+
+You can compare new way of handling comtime to Zig comptime functionality. It would allow us not only to perform some computations at compile-time but also generate new structures and functions at compile-time. All generic types would not be just some compiler magic but `v.ast.Type` and you should be able to invoke all of its methods at compile-time, get type symbol from it and even create new `v.ast.Type`s at compile-time.  
+
+First where comptime can be used is automatic generation of serialization/deserialization functions. Here's small example of how we can implement automatic `str()` for types: 
+```v
+fn to_str<T>(val T) string {
+    // FIXME: Decide if generic type is `v.ast.Type` or `v.ast.TypeSymbol`
+    $if T.kind == .int {
+        return int_string(val)
+    } else if T.kind == .array {
+        mut out := '['
+        for i,elem in val {
+            out = '${out}${elem}'
+            if i != val.len - 1 {
+                out = '${out},'
+            }
+        }
+        return '${out}]'
+    } else if T.kind == .struct_ {
+        out = '${$T.name} {'
+        $for field in T.fields {
+            field := val.$field 
+            out = '${out} ${$field}: ${field}'
+        }
+        return '${out}}'
+    } else {
+        // and so on...
+    }
+}
+```
+
+Notice that we do not have `$else` anymore, that is because with this proposal if parser sees `$` it simply continues parsing V statements or expressions *but* parsed statement and expression will be marked with `comptime` flag.
+
+And here is simple example of factorial that is executed fully at compile-time! 
+```v
+fn factorial(x int) int {
+    if x < 2 {
+        return x 
+    } 
+
+    return factorial(x - 1) * x
+}
+
+fn main() {
+    cx := $factorial(5) // call prefixed with `$` becomes compile-time call
+    $cy := factorial(6) // variable assignment prefixed with `$` puts it to compile-time environment 
+    println(cx) // 120 
+    println($cy) // 720
+}
+```
```

Comment:
This needs further more detailed explanation, about what will be produced at each stage: what will be printed, and *when* it will be printed.